### ISSUE TITLE
Probability floating point error

### ISSUE THREAD

**Issue by tommycarstensen** [genetics-statistics/GEMMA#130]
I think there might be a bug in GEMMA. When the probability is less than 10^-308, then 0.000000e+00 is printed instead. I think it might be a floating point error:
https://en.wikipedia.org/wiki/IEEE_754#Basic_and_interchange_formats

**Comment by pjotrp** [comment]
Well we do round stuff. Do you have a reproducible example for us?

**Comment by pcarbo** [comment]
@tommycarstensen Probability or *p*-value? How do you know it is less than `10^(-308)`?

**Comment by tommycarstensen** [comment]
@pjotrp This is not a rounding issue. The output is wrong and my collaborator didn't pick up on the missing signal.

@pcarbo I am assuming it is less than 10^-308, because I got 10^-306 with slightly different settings (a different kinship matrix not using LOCO) and the Decimal E max for binary64 is 307.95:
https://en.wikipedia.org/wiki/Double-precision_floating-point_format

I can generate a reproducible example for you, but I'm extremely busy at the moment. Please let me know, whether you need a minimal working example to troubleshoot this or not. I will provide you with one in 2-3 weeks, if you need it to fix this bug. Thanks guys!

**Comment by pjotrp** [comment]
It is a LOCO thing? Can you send us a command line you used? If I can't replicate it I may ask for data next.

**Comment by pcarbo** [comment]
@pjotrp Do you know if GEMMA is storing the *p*-values or the log(*p*-values) in doubles? If it is storing the p-values, then the error would be due to overflow.

**Comment by tommycarstensen** [comment]
@pjotrp Let me know, if you need the data. I can prepare it. I'm quite sure it's a floating point issue. I'm not familiar with your codebase, but as suggested by @pcarbo it's probably because the p-values are not stored as their log values.

**Comment by pjotrp** [comment]
Having data makes it much easier to fix. So, yes please.

**Comment by pcarbo** [comment]
@tommycarstensen We would also need the exact command that you ran. We would only need the genotypes for the one SNP, and the relatedness matrix if you are using the LMM.

**Comment by tommycarstensen** [comment]
It might be easier for me to identify the problem in your code base instead of preparing and uploading the data. Let me check your code and get back to you.

**Comment by tommycarstensen** [comment]
Here is the command:
gemma0.94.1 -lmm 4 -g $g -p $p -k $k -o $o -notsnp -n 1 -c $c -r2 1

Here are the files:
p=[http://tommycarstensen.com/gemma.130.p.txt](http://tommycarstensen.com/gemma.130.p.txt)
g=[http://tommycarstensen.com/gemma.130.g.bimbam.gz](http://tommycarstensen.com/gemma.130.g.bimbam.gz)
k=[http://tommycarstensen.com/gemma.130.k.cXX.txt](http://tommycarstensen.com/gemma.130.k.cXX.txt)

You should also get the `0.000000e+00` printed without the `-c` flag.

Here a few faulty lines:

```
    -9	2:234664586	-9	0	ATC	A	0.547	-6.727135e-01	1.620261e-02	4.116354e+00	4.114995e+00	0.000000e+00	0.000000e+00	4.987614e-265
    -9	2:234665983	-9	0	A	G	0.543	-6.747418e-01	1.622826e-02	4.076411e+00	4.075001e+00	0.000000e+00	0.000000e+00	2.117264e-265
    -9	2:234668570	-9	0	C	T	0.542	-6.690000e-01	1.615895e-02	4.058750e+00	4.057324e+00	0.000000e+00	0.000000e+00	1.165549e-263
    -9	2:234673309	-9	0	C	T	0.544	-6.677777e-01	1.619654e-02	4.140583e+00	4.139261e+00	0.000000e+00	0.000000e+00	1.967984e-262
```

I hope you find the bug and fix it. Best of luck. Kindly let me know, if I can provide additional details.

Feature request: Can you make GEMMA print the command that was used to stdout or stderr. That would make troubleshooting like this easier and would be good practice. Thanks!

**Comment by pjotrp** [comment]
That is a large dataset and I ran a recent version of gemma which actually bails out:

```sh
gemma -lmm 4 -g gemma.130.g.bimbam.gz -p gemma.130.p.txt -k gemma.130.k.cXX.txt -notsnp -n 1 -r2 1 -debug     
```

```
GEMMA 0.97 (2017/12/19) by Xiang Zhou and team (C) 2012-2017
Reading Files ... 
**** DEBUG: entered in src/io.cpp at line 353 in ReadFile_pheno
**** DEBUG: entered in src/io.cpp at line 608 in ReadFile_geno
**** DEBUG: Can't figure out position for 2:233649290 in src/io.cpp at line 683 in ReadFile_geno
(...)
**** DEBUG: Skipping similar warnings in src/io.cpp at line 685 in ReadFile_geno
## number of total individuals = 6407
## number of analyzed individuals = 6407
## number of covariates = 1
## number of phenotypes = 1
## number of total SNPs/var        =   100000
## number of analyzed SNPs         =   100000
**** DEBUG: entered in src/io.cpp at line 1126 in ReadFile_kin
Start Eigen-Decomposition...
pve estimate =0.43298
se(pve) =0.0287056
**** DEBUG: gemma.130.g.bimbam.gz in src/lmm.cpp at line 1467 in AnalyzeBimbam
ERROR: Enforce failed for N>0 in src/fastblas.cpp at line 194 in fast_cblas_dgemm
```

Looks like we have another integer overflow error with BLAS, similar to #120 which had 6,000 individuals. Good thing we are enforcing a check these days.

Gemma normally outputs a log file with the command line. But since this program bailed out it was not written.

For now reduce the number of individuals. When I have time I'll look into fixing it. Thanks for reporting.

**Comment by pjotrp** [comment]
it also bails out in -legacy mode.

**Comment by tommycarstensen** [comment]
@pjotrp Thanks! Just a note. The older `version 0.94.1` runs "successfully" to completion. It prints what seems to be correct p values for the remainder of the variants in there. It's only when the p value is less than `10^-308` that `0.000000e+00` seems to be printed instead.

P.S. Sorry it took so long to upload example files. I've been really busy.

**Comment by pjotrp** [comment]
Thanks. Needs fixing for sure.

**Comment by pcarbo** [comment]
@pjotrp Here's a phenotype file you can use for testing because it only has 11 SNPs (so you don't have to worry about the genotype matrix being too large); the 6th SNP reproduces the same error that @tommycarstensen observed above.

[gemma.130.p.txt.gz](https://github.com/genetics-statistics/GEMMA/files/1738885/gemma.130.p.txt.gz)

**Comment by pcarbo** [comment]
@pjotrp I was able to reproduce the same error in GEMMA 0.95a on my MacBook Pro.

Some more interesting observations when running GEMMA version 0.95a on this data set (the one with 11 SNPs). Interestingly, I found that reading the data took a few seconds in 0.95a, and *much* longer in 0.97; the whole analysis took 80 seconds in GEMMA 0.95a (on my MacBook Pro), and 224 seconds in GEMMA 0.97. 

Any idea what the new version is so much slower at loading the data?

**Comment by pjotrp** [comment]
@pcarbo can you provide gemma.130.g.small.bimbam?

**Comment by pcarbo** [comment]
@pjotrp You are right, I attached the wrong file. Please see [above](https://github.com/genetics-statistics/GEMMA/issues/130#issuecomment-366862441).

**Comment by pjotrp** [comment]
Thanks, now I can confirm both this issue and the speed regression#136. Thanks, this makes it likely we can fix both.

**Comment by pjotrp** [comment]
Fixed,  see https://github.com/genetics-statistics/GEMMA/issues/182. Turns out a rounded number of SNPs gave the problem.
